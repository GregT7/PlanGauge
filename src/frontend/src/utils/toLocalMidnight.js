export default function toLocalMidnight(date) {
  if (typeof date === "string") {
    // manually parse as local date instead of letting Date guess UTC
    const [year, month, day] = date.split("-").map(Number);
    return new Date(year, month - 1, day); // Local midnight
  } else if (date instanceof Date) {
    // Assume already a local date (from picker or programmatic input)
    return new Date(date.getFullYear(), date.getMonth(), date.getDate());
  }
  return null; // or throw an error
}
