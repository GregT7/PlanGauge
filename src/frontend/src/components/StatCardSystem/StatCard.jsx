import {
  Card,
  CardAction,
  CardContent,
  CardDescription,
  CardFooter,
  CardHeader,
  CardTitle,
} from "@/components/ui/card"
import toLocalMidnight from "@/utils/toLocalMidnight";
import { format } from "date-fns";

function capitalizeFirstLetter(str) {
  if (typeof str !== 'string' || str.length === 0) {
    return str; // Handle non-string or empty inputs
  }
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function determineStatusStyle(status) {
    switch (status) {
        case "neutral":
            return "bg-zinc-900";
        case "good":
            return "bg-emerald-900 border-2 border-emerald-400"
        case "moderate":
            return "bg-amber-900 border-2 border-amber-400"
        case "poor":
            return "bg-rose-900 border-2 border-rose-400";
        default:
            return "bg-red-900 border-2 border-red-500"
    }
}

const StatCard = ({cardData, className=""}) => {
    const cardDate = toLocalMidnight(cardData.date);
    const desciptStr = format(cardDate, 'MMMM d, yyyy')
    
    return (
        <Card className={`${className} ${determineStatusStyle(cardData.status)}`}>
            <CardHeader>
                <CardTitle>{cardData.name}</CardTitle>
                <CardDescription>{desciptStr}</CardDescription>
            </CardHeader>
            <CardContent>
                <p>{`Status: ${capitalizeFirstLetter(cardData.status)}`}</p>
                <p>{`Sum: ${cardData.sum} mins`}</p>
            </CardContent>
        </Card>
    );
}

export default StatCard;