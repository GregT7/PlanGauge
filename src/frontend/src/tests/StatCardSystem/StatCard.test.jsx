import { expect, it, describe } from "vitest";
import { screen, render } from "@testing-library/react";
import StatCard from "@/components/StatCardSystem/StatCard";
import toLocalMidnight from "@/utils/toLocalMidnight";
import { format } from "date-fns";

const cardData = { "name": "Monday", "ave": 90, "std": 35, "date": "2025-06-16", "status": "neutral", "sum": 500 };
const goodData = { "name": "Tuesday", "ave": 90, "std": 35, "date": "2025-06-17", "status": "good", "sum": 90 };
const moderateData = { "name": "Wednesday", "ave": 90, "std": 35, "date": "2025-06-18", "status": "moderate", "sum": 990 };
const poorData = { "name": "Thursday", "ave": 90, "std": 35, "date": "2025-06-19", "status": "poor", "sum": 34 };
const unknownData = { "name": "Monday", "ave": 90, "std": 35, "date": "2025-06-16", "status": "unknown", "sum": -1 }

describe("StatCard unit testing", () => {
    it("passes snapshot test", () => {
        const { container } = render(<StatCard cardData={cardData}/>);
        expect(container).toMatchSnapshot();
    })

    it("handles empty 'cardData' and 'className' props passed", ()=> {
        render(<StatCard/>)
        expect(screen.getByTestId('StatCard')).toBeInTheDocument();
    })

    it("handles invalid className prop", () => {
        render(<StatCard className={{test: "123"}}/>)
        expect(screen.getByTestId('StatCard')).toBeInTheDocument();
    });

    // completely invalid: not an object, or an object with no matching keys
    it("handles completely invalid 'cardData' passed as props", () => {
        const cardData = "data 123"
        render(<StatCard cardData={cardData}/>)
        expect(screen.getByTestId('StatCard')).toBeInTheDocument();
    })

    //  { "name": "Monday",    "ave": 90, "std": 35, "date": "2025-06-16", "status": "neutral", "sum": 0 },
    it("correctly parses & renders cardData when valid props are passed", () => {
        render(<StatCard cardData={cardData}/>);

        const dateStr = format(toLocalMidnight(cardData.date), 'MMMM d, yyyy')
        expect(screen.getByText(dateStr)).toBeInTheDocument();
        expect(screen.getByText("Monday")).toBeInTheDocument();

        let capitalizedStr = cardData.status.charAt(0).toUpperCase() + cardData.status.slice(1)
        capitalizedStr = "Status: " + capitalizedStr;
        expect(screen.getByText(capitalizedStr)).toBeInTheDocument();

        const sumText = "Sum: " + cardData.sum + " mins"
        expect(screen.getByText(sumText)).toBeInTheDocument();


        screen.debug();
    })



    it("applies the correct styling based on passed status (neutral, good, moderate, poor, uknown)", () => {
        // status: neutral
        const { unmount: unmount1 } = render(<StatCard cardData={cardData}/>)
        expect(screen.getByTestId("StatCard")).toHaveClass("bg-zinc-900");
        unmount1();

        // status: good
        const { unmount: unmount2 } = render(<StatCard cardData={goodData}/>)
        expect(screen.getByTestId("StatCard")).toHaveClass("bg-emerald-800 border-emerald-600");
        unmount2();

        // status: moderate
        const { unmount: unmount3 } = render(<StatCard cardData={moderateData}/>)
        expect(screen.getByTestId("StatCard")).toHaveClass("bg-amber-800 border-2 border-amber-600");
        unmount3();

        // status: poor
        const { unmount: unmount4 } = render(<StatCard cardData={poorData}/>)
        expect(screen.getByTestId("StatCard")).toHaveClass("bg-rose-800 border-2 border-rose-600");
        unmount4();

        // status: unknown
        const { unmount: unmount5 } = render(<StatCard cardData={unknownData}/>)
        expect(screen.getByTestId("StatCard")).toHaveClass("bg-red-800 border-2 border-red-600");
        unmount5();
    })
})