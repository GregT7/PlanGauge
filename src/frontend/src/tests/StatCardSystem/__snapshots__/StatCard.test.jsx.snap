// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StatCard unit testing > passes snapshot test 1`] = `
<div>
  <div
    class="text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm bg-zinc-900 h-full overflow-hidden"
    data-slot="card"
    data-testid="StatCard"
  >
    <div
      class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 space-y-1"
      data-slot="card-header"
    >
      <div
        class="font-semibold text-base truncate"
        data-slot="card-title"
      >
        Monday
      </div>
      <div
        class="text-muted-foreground text-sm truncate"
        data-slot="card-description"
      >
        June 16, 2025
      </div>
    </div>
    <div
      class="px-6 space-y-1 text-sm"
      data-slot="card-content"
    >
      <p
        class="truncate"
      >
        Status: Neutral
      </p>
      <p
        class="truncate"
      >
        Sum: 500 mins
      </p>
    </div>
  </div>
</div>
`;
